File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
        ']':'[',
        '}':'{',
        ')':'('
        
        }
        
        for char in s:
            if char in ['(','[','{']:
                stack.append(char)
            else:
                if stack and stack[-1] == pairs[char]:
                    stack.pop()
                else:
                    return False
        if len(stack) > 0:
            return False
        else:
            return True

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack  = new Stack<>();

        Map<Character, Character> brackets = new HashMap<>();
        brackets.put('}', '{');
        brackets.put(']', '[');
        brackets.put(')', '(');

        List<Character> openBrackets = List.of('{', '[', '(');

        for(int i = 0; i < s.length(); i++){
            char current = s.charAt(i);
            if(openBrackets.contains(current)){
                stack.add(current);
            } else {
                if (stack.isEmpty() || stack.pop() != brackets.get(current)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}